Mathlib/AlgebraicGeometry/EllipticCurve/NormalForms.lean

```diff
@@ -0,0 +1,658 @@
+
+end IsCharThreeJNeZeroNF
+
+/-! ### Normal forms of characteristic = 3 -/
+
+/-- A `WeierstrassCurve` is in normal form of characteristic = 3, if it is
+$Y^2 = X^3 + a_2X^2 + a_6$ (`WeierstrassCurve.IsCharThreeJNeZeroNF`) or
+$Y^2 = X^3 + a_4X + a_6$ (`WeierstrassCurve.IsShortNF`). -/
+def IsCharThreeNF : Prop := W.IsCharThreeJNeZeroNF ∨ W.IsShortNF
+
+namespace IsCharThreeNF
+
+variable {W} (self : W.IsCharThreeNF)
+include self
+
+theorem isCharNeTwoNF : W.IsCharNeTwoNF :=
+  match self with
+  | Or.inl h | Or.inr h => h.isCharNeTwoNF
+
+end IsCharThreeNF
+
+section VariableChange
+
+variable [CharP R 3] [CharP F 3]
+
+/-- For a `WeierstrassCurve` defined over a ring of characteristic = 3,
+there is an explicit change of variables, which changes it to $Y^2 = X^3 + a_4X + a_6$
+(`WeierstrassCurve.IsShortNF`) if its j = 0.
+This is in fact given by `WeierstrassCurve.toCharNeTwoNF`. -/
+def toShortNFOfCharThree : VariableChange R :=
+  have h : (2 : R) * 2 = 1 := by linear_combination CharP.cast_eq_zero R 3
+  letI : Invertible (2 : R) := ⟨2, h, h⟩
+  W.toCharNeTwoNF
+
+/-- For a `WeierstrassCurve` defined over a field of characteristic = 3,
+there is an explicit change of variables of it to `WeierstrassCurve.IsCharThreeNF`, that is,
+$Y^2 = X^3 + a_2X^2 + a_6$ (`WeierstrassCurve.IsCharThreeJNeZeroNF`) or
+$Y^2 = X^3 + a_4X + a_6$ (`WeierstrassCurve.IsShortNF`).
+It is the composition of an explicit change of variables with
+`WeierstrassCurve.toShortNFOfCharThree`. -/
+def toCharThreeNF (W : WeierstrassCurve F) : VariableChange F :=
+  ⟨1, (W.variableChange W.toShortNFOfCharThree).a₄ /
+    (W.variableChange W.toShortNFOfCharThree).a₂, 0, 0⟩ * W.toShortNFOfCharThree
+
+lemma toShortNFOfCharThree_a₂ : (W.variableChange W.toShortNFOfCharThree).a₂ = W.b₂ := by
+  simp_rw [toShortNFOfCharThree, toCharNeTwoNF, variableChange_a₂, inv_one, Units.val_one, b₂]
+  linear_combination (-W.a₂ - W.a₁ ^ 2) * CharP.cast_eq_zero R 3
+
+theorem toShortNFOfCharThree_spec (hb₂ : W.b₂ = 0) :
+    (W.variableChange W.toShortNFOfCharThree).IsShortNF := by
+  have h : (2 : R) * 2 = 1 := by linear_combination CharP.cast_eq_zero R 3
+  letI : Invertible (2 : R) := ⟨2, h, h⟩
+  have H := W.toCharNeTwoNF_spec
+  exact ⟨H.a₁, hb₂ ▸ W.toShortNFOfCharThree_a₂, H.a₃⟩
+
+section
+
+variable (W : WeierstrassCurve F)
+
+theorem toCharThreeNF_spec_of_b₂_ne_zero (hb₂ : W.b₂ ≠ 0) :
+    (W.variableChange W.toCharThreeNF).IsCharThreeJNeZeroNF := by
+  have h : (2 : F) * 2 = 1 := by linear_combination CharP.cast_eq_zero F 3
+  letI : Invertible (2 : F) := ⟨2, h, h⟩
+  rw [toCharThreeNF]; erw [variableChange_comp]
+  rw [← toShortNFOfCharThree_a₂] at hb₂
+  set W' := W.variableChange W.toShortNFOfCharThree
+  have H : W'.IsCharNeTwoNF := W.toCharNeTwoNF_spec
+  constructor
+  · simp [H.a₁]
+  · simp [H.a₁, H.a₃]
+  · field_simp
+    linear_combination (W'.a₄ * W'.a₂ ^ 2 + W'.a₄ ^ 2) * CharP.cast_eq_zero F 3
+
+theorem toCharThreeNF_spec_of_b₂_zero (hb₂ : W.b₂ = 0) :
```

Comment:
```suggestion
theorem toCharThreeNF_spec_of_b₂_eq_zero (hb₂ : W.b₂ = 0) :
```

Reply:
But do you think we need to rename `XXXJZero -> XXXJEqZero`? In my opinion, `JEqZero` is too long...

Reply:
Yeah I'm also a bit torn about this. I prefer changing `JZero` to `JEqZero` because `J` isn't a keyword used in other places in mathlib, unlike `Char` which is standard. I'm happy with either, but this lemma can be `_eq_zero` regardless.

Reply:
Changed.